File: web/src/stats.ts
/**
 * Measuring how converged the render is, on the GPU. Mirrors
 * `crates/gpu/src/stats.rs`; the arithmetic lives in
 * `shaders/stats/convergence.wgsl`.
 *
 * # Why measure instead of modelling
 *
 * `1/sqrt(N)` is one line of TypeScript and needs no GPU work at all. It is also
 * wrong exactly when someone is looking at the number: it describes an estimator
 * with finite, well-sampled variance, and the scenes where anyone wonders "is
 * this done yet" are the ones with a caustic or a firefly making that false.
 * Measured against sixteen independent renders, this reduction lands within 4%
 * of the true spread; the model is off by whatever the scene decides.
 *
 * # Why it never blocks
 *
 * A `mapAsync` awaited inside the frame loop would serialise the CPU against the
 * GPU and turn a diagnostic into a frame-rate cost. Instead the readback runs on
 * a small pool of staging buffers, the result lands whenever it lands, and the
 * displayed figure is simply a frame or two old — which for a number that moves
 * on the scale of seconds is not a difference anyone can perceive.
 */

/**
 * Workgroups in the first pass.
 *
 * Fixed rather than derived from the pixel count, so the partials buffer and the
 * second pass's fold are both constant-sized. At the shader's workgroup size of
 * 256 that is 16 384 threads, which saturates any GPU this runs on; each one
 * strides through the image rather than taking a contiguous block, so the loads
 * coalesce.
 */
const GROUPS = 64;

const PARAMS_BYTES = 16;

/** Staging buffers in flight at once. */
const POOL = 3;

export class ConvergenceMeter {
  private partials: GPUBuffer;
  private result: GPUBuffer;
  private params: GPUBuffer;

  /** Unmapped staging buffers, available to receive a copy. */
  private free: GPUBuffer[] = [];
  /** Copied into this frame, waiting for a `mapAsync` that submit must precede. */
  private queued: GPUBuffer[] = [];

  /**
   * The most recent measurement: mean relative standard error over the pixels
   * bright enough to have one. Negative means "nothing measured yet", which the
   * UI shows as a dash rather than as a confident zero.
   */
  value = -1;

  private constructor(
    private device: GPUDevice,
    private layout: GPUBindGroupLayout,
    private reduce: GPUComputePipeline,
    private finish: GPUComputePipeline,
  ) {
    this.partials = device.createBuffer({
      label: 'convergence partials',
      size: GROUPS * 8,
      usage: GPUBufferUsage.STORAGE,
    });
    this.result = device.createBuffer({
      label: 'convergence result',
      size: 4,
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
    });
    this.params = device.createBuffer({
      label: 'convergence params',
      size: PARAMS_BYTES,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });
    for (let i = 0; i < POOL; i++) {
      this.free.push(
        device.createBuffer({
          label: `convergence readback ${i}`,
          size: 4,
          usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
        }),
      );
    }
  }

  static async create(device: GPUDevice, module: GPUShaderModule): Promise<ConvergenceMeter> {
    const sto = (binding: number, readOnly: boolean): GPUBindGroupLayoutEntry => ({
      binding,
      visibility: GPUShaderStage.COMPUTE,
      buffer: { type: readOnly ? 'read-only-storage' : 'storage' },
    });
    const layout = device.createBindGroupLayout({
      label: 'convergence',
      entries: [
        { binding: 0, visibility: GPUShaderStage.COMPUTE, buffer: { type: 'uniform' } },
        sto(1, true),
        sto(2, false),
        sto(3, false),
      ],
    });
    const pl = device.createPipelineLayout({ bindGroupLayouts: [layout] });
    const stage = (entryPoint: string) =>
      device.createComputePipelineAsync({
        label: entryPoint,
        layout: pl,
        compute: { module, entryPoint },
      });
    const [reduce, finish] = await Promise.all([stage('reduce'), stage('finish')]);
    return new ConvergenceMeter(device, layout, reduce, finish);
  }

  /**
   * Record the reduction into `encoder` and copy the answer to a staging buffer.
   *
   * Returns false, having recorded nothing, when every staging buffer is still
   * in flight. Dropping a measurement is the right response to that: the figure
   * is a readout, and one taken a frame later says the same thing.
   */
  record(encoder: GPUCommandEncoder, accum: GPUBuffer, pixels: number): boolean {
    const staging = this.free.pop();
    if (!staging) return false;

    const p = new Uint32Array([pixels, GROUPS, 0, 0]);
    this.device.queue.writeBuffer(this.params, 0, p);

    const group = this.device.createBindGroup({
      label: 'convergence',
      layout: this.layout,
      entries: [
        { binding: 0, resource: { buffer: this.params } },
        { binding: 1, resource: { buffer: accum } },
        { binding: 2, resource: { buffer: this.partials } },
        { binding: 3, resource: { buffer: this.result } },
      ],
    });

    // Two passes, and they must be separate: the fold reads every partial the
    // first pass wrote, and a `storageBarrier` only synchronises within a
    // workgroup. Ending the pass is the only barrier WebGPU gives across them.
    for (const [pipeline, groups] of [
      [this.reduce, GROUPS],
      [this.finish, 1],
    ] as const) {
      const pass = encoder.beginComputePass({ label: 'convergence' });
      pass.setPipeline(pipeline);
      pass.setBindGroup(0, group);
      pass.dispatchWorkgroups(groups, 1, 1);
      pass.end();
    }
    encoder.copyBufferToBuffer(this.result, 0, staging, 0, 4);
    this.queued.push(staging);
    return true;
  }

  /**
   * Start the readbacks for whatever `record` queued. Call after `submit`.
   *
   * `mapAsync` on a buffer whose copy has not been submitted is a validation
   * error, so this cannot be folded into `record`.
   */
  collect(): void {
    const queued = this.queued;
    this.queued = [];
    for (const staging of queued) {
      staging
        .mapAsync(GPUMapMode.READ)
        .then(() => {
          this.value = new Float32Array(staging.getMappedRange())[0];
          staging.unmap();
          this.free.push(staging);
        })
        .catch(() => {
          // A destroyed device rejects every pending map. Drop the buffer rather
          // than returning it to the pool, so a lost device does not spin here.
        });
    }
  }

  destroy(): void {
    for (const b of [this.partials, this.result, this.params, ...this.free]) b.destroy();
    this.free = [];
  }
}

/**
 * Samples still needed to reach `target` relative error, at the current rate.
 *
 * Error falls as `1/sqrt(N)`, so `N_target = N * (e / e_target)^2`. This is the
 * one place a model is the right tool: it extrapolates from a *measured* point
 * rather than replacing the measurement, and the user gets a fresh measurement
 * every few frames to correct it.
 */
export function samplesToReach(current: number, samples: number, target: number): number {
  if (current <= 0 || samples <= 0 || target <= 0) return 0;
  return Math.max(0, Math.ceil(samples * (current / target) ** 2) - samples);
}

File: crates/gpu/src/stats.rs
//! Measuring how converged a render is, on the GPU.
//!
//! See `shaders/stats/convergence.wgsl` for what the number means. The host side
//! is a two-pass reduction and one four-byte readback.

use crate::{storage_buffer, Gpu, GpuError};
use bytemuck::{Pod, Zeroable};
use wgpu::util::DeviceExt;

#[repr(C)]
#[derive(Clone, Copy, Pod, Zeroable, Default)]
struct StatsParams {
    pixels: u32,
    partials: u32,
    _pad0: u32,
    _pad1: u32,
}

/// Workgroups in the first pass.
///
/// Fixed rather than derived from the pixel count: it bounds the partials buffer
/// and the second pass's loop, and 64 groups of the shader's 256 threads
/// saturate any GPU this runs on while keeping the fold trivially small.
const GROUPS: u32 = 64;

/// Compiled reduction kernels.
pub struct ConvergenceMeter {
    layout: wgpu::BindGroupLayout,
    reduce: wgpu::ComputePipeline,
    finish: wgpu::ComputePipeline,
    partials: wgpu::Buffer,
    result: wgpu::Buffer,
    staging: wgpu::Buffer,
}

impl ConvergenceMeter {
    pub fn new(gpu: &Gpu) -> Result<Self, GpuError> {
        let device = &gpu.device;
        let module = gpu.create_shader("stats/convergence.wgsl")?;
        let sto = |b: u32, read_only: bool| wgpu::BindGroupLayoutEntry {
            binding: b,
            visibility: wgpu::ShaderStages::COMPUTE,
            ty: wgpu::BindingType::Buffer {
                ty: wgpu::BufferBindingType::Storage { read_only },
                has_dynamic_offset: false,
                min_binding_size: None,
            },
            count: None,
        };
        let layout = device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
            label: Some("convergence"),
            entries: &[
                wgpu::BindGroupLayoutEntry {
                    binding: 0,
                    visibility: wgpu::ShaderStages::COMPUTE,
                    ty: wgpu::BindingType::Buffer {
                        ty: wgpu::BufferBindingType::Uniform,
                        has_dynamic_offset: false,
                        min_binding_size: None,
                    },
                    count: None,
                },
                sto(1, true),
                sto(2, false),
                sto(3, false),
            ],
        });
        let pl = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
            label: Some("convergence"),
            bind_group_layouts: &[Some(&layout)],
            ..Default::default()
        });
        let stage = |entry_point: &str| {
            device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
                label: Some(entry_point),
                layout: Some(&pl),
                module: &module,
                entry_point: Some(entry_point),
                compilation_options: Default::default(),
                cache: None,
            })
        };
        let zeros = vec![0.0f32; (GROUPS * 2) as usize];
        Ok(Self {
            layout,
            reduce: stage("reduce"),
            finish: stage("finish"),
            partials: storage_buffer(device, "partials", bytemuck::cast_slice(&zeros), 8),
            result: device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some("convergence result"),
                contents: bytemuck::bytes_of(&-1.0f32),
                usage: wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::COPY_SRC,
            }),
            staging: device.create_buffer(&wgpu::BufferDescriptor {
                label: Some("convergence readback"),
                size: 4,
                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
                mapped_at_creation: false,
            }),
        })
    }

    /// Record the reduction and copy the result to the staging buffer.
    pub fn record(&self, gpu: &Gpu, encoder: &mut wgpu::CommandEncoder, accum: &wgpu::Buffer, pixels: u32) {
        let params = gpu
            .device
            .create_buffer_init(&wgpu::util::BufferInitDescriptor {
                label: Some("stats params"),
                contents: bytemuck::bytes_of(&StatsParams {
                    pixels,
                    partials: GROUPS,
                    _pad0: 0,
                    _pad1: 0,
                }),
                usage: wgpu::BufferUsages::UNIFORM,
            });
        let group = gpu.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("convergence"),
            layout: &self.layout,
            entries: &[
                wgpu::BindGroupEntry {
                    binding: 0,
                    resource: params.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 1,
                    resource: accum.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 2,
                    resource: self.partials.as_entire_binding(),
                },
                wgpu::BindGroupEntry {
                    binding: 3,
                    resource: self.result.as_entire_binding(),
                },
            ],
        });
        for (pipeline, groups) in [(&self.reduce, GROUPS), (&self.finish, 1)] {
            let mut p = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
                label: Some("convergence"),
                timestamp_writes: None,
            });
            p.set_pipeline(pipeline);
            p.set_bind_group(0, &group, &[]);
            p.dispatch_workgroups(groups, 1, 1);
        }
        encoder.copy_buffer_to_buffer(&self.result, 0, &self.staging, 0, 4);
    }

    /// Read the recorded result. Blocks; the browser reads it asynchronously
    /// instead so a stat never stalls a frame.
    pub fn read(&self, gpu: &Gpu) -> Result<f32, GpuError> {
        let slice = self.staging.slice(..);
        let (tx, rx) = std::sync::mpsc::channel();
        slice.map_async(wgpu::MapMode::Read, move |r| {
            let _ = tx.send(r);
        });
        gpu.device
            .poll(wgpu::PollType::wait_indefinitely())
            .map_err(|e| GpuError::Readback(format!("device poll: {e}")))?;
        rx.recv()
            .map_err(|e| GpuError::Readback(e.to_string()))?
            .map_err(|e| GpuError::Readback(e.to_string()))?;
        let data = slice
            .get_mapped_range()
            .map_err(|e| GpuError::Readback(e.to_string()))?;
        let v = f32::from_ne_bytes([data[0], data[1], data[2], data[3]]);
        drop(data);
        self.staging.unmap();
        Ok(v)
    }
}
